src/Chessboard.cpp

```diff
@@ -0,0 +1,313 @@
+#include "Chessboard.h"
+#include "Rook.h"
+#include "Bishop.h"
+#include "King.h"
+#include "Queen.h"
+#include "Knight.h"
+#include "Pawn.h"
+
+Chessboard::Chessboard() {
+    initializeBoard();
+    currentPlayer_ = PieceColor::White;
+}
+
+
+Chessboard::~Chessboard() {
+    clearBoard();
+}
+
+void Chessboard::initializeBoard() {
+    for (int row = 0; row < 8; row++) {
+        for (int col = 0; col < 8; col++) {
+            board_[row][col] = nullptr;
+        }
+    }
+
+    // initialize white pieces
+    board_[0][0] = new Rook(PieceColor::White, 0, 0);
+    board_[0][1] = new Knight(PieceColor::White, 0, 1);
+    board_[0][2] = new Bishop(PieceColor::White, 0, 2);
+    board_[0][3] = new Queen(PieceColor::White, 0, 3);
+    board_[0][4] = new King(PieceColor::White, 0, 4);
+    board_[0][5] = new Bishop(PieceColor::White, 0, 5);
+    board_[0][6] = new Knight(PieceColor::White, 0, 6);
+    board_[0][7] = new Rook(PieceColor::White, 0, 7);
+
+    for (int i = 0; i < 8; i++) {
+        board_[1][i] = new Pawn(PieceColor::White, 1, i);
+    }
+
+    // initialize black pieces
+    board_[7][0] = new Rook(PieceColor::Black, 7, 0);
+    board_[7][1] = new Knight(PieceColor::Black, 7, 1);
+    board_[7][2] = new Bishop(PieceColor::Black, 7, 2);
+    board_[7][3] = new Queen(PieceColor::Black, 7, 3);
+    board_[7][4] = new King(PieceColor::Black, 7, 4);
+    board_[7][5] = new Bishop(PieceColor::Black, 7, 5);
+    board_[7][6] = new Knight(PieceColor::Black, 7, 6);
+    board_[7][7] = new Rook(PieceColor::Black, 7, 7);
+
+    for (int i = 0; i < 8; i++) {
+        board_[6][i] = new Pawn(PieceColor::Black, 6, i);
+    }
+}
+
+
+void Chessboard::clearBoard() {
+    for (int i = 0; i < 8; i++) {
+        for (int j = 0; j < 8; j++) {
+            if (board_[i][j] != nullptr) {
+                delete board_[i][j];
+            }
+        }
+    }
+}
+
+int Chessboard::movePiece(int startRow, int startCol, int endRow, int endCol) {
+    Piece* startPiece = board_[startRow][startCol];
+    Piece* endPiece = board_[endRow][endCol];
+
+    if (startPiece == nullptr) {
+        return 11; // no piece to move
+    }
+
+    if (startPiece->getColor() != currentPlayer_) {
+        return 12; // can't move opponent's piece
+    }
+
+    if (endPiece != nullptr && endPiece->getColor() == currentPlayer_) {
+        return 13; // can't capture own piece
+    }
+
+    if (!startPiece->isValidMove(endRow, endCol)) {
+        return 21; // invalid move
+    }
+
+    Chessboard copy = *this;
+    copy.board_[endRow][endCol] = startPiece;
+    copy.board_[startRow][startCol] = nullptr;
+    startPiece->setRow(endRow);
+    startPiece->setCol(endCol);
+
+    if (copy.isCheck(currentPlayer_)) {
+        return 31; // moving will cause checkmate
+    }
+
+    board_[endRow][endCol] = startPiece;
+    board_[startRow][startCol] = nullptr;
+    startPiece->setRow(endRow);
+    startPiece->setCol(endCol);
+    if (isCheck(currentPlayer_)) {

```

Comment:
You can simplify the logic by checking for checkmate only once after the move is made. If the move causes a checkmate, you can immediately return the corresponding error code.